File: 0118-pascals-triangle/0118-pascals-triangle.c
#include <stdlib.h>

/**
 * @brief Generate Pascal's Triangle up to 'numRows' rows.
 *
 * @param numRows (int): The number of rows of Pascal's Triangle to generate.
 * @param returnSize (int*): Pointer to return the number of rows (numRows).
 * @param returnColumnSizes (int**): Pointer to return the size of each row (number of elements in each row).
 *
 * @return (int**): A dynamically allocated 2D array containing Pascal's Triangle.
 *                  The caller must free both the returned array and the returnColumnSizes array.
 */
int** generate(int numRows, int* returnSize, int** returnColumnSizes) {
    // Allocate memory for the Pascal's Triangle (2D array)
    int** pascalTriangle = (int**)malloc(numRows * sizeof(int*));

    // Allocate memory for the returnColumnSizes array to store the size of each row
    *returnColumnSizes = (int*)malloc(numRows * sizeof(int));

    // Generate each row of Pascal's Triangle
    for (int i = 0; i < numRows; i++) {
        // Allocate memory for the current row (i + 1 elements)
        pascalTriangle[i] = (int*)malloc((i + 1) * sizeof(int));

        // Set the size of the current row in the returnColumnSizes array
        (*returnColumnSizes)[i] = i + 1;

        // First and last elements of the row are always 1
        pascalTriangle[i][0] = 1;
        pascalTriangle[i][i] = 1;

        // Fill the elements between the first and last by summing values from the previous row
        for (int j = 1; j < i; j++) {
            pascalTriangle[i][j] = pascalTriangle[i - 1][j - 1] + pascalTriangle[i - 1][j];
        }
    }

    // Set the return size (number of rows)
    *returnSize = numRows;

    // Return the generated Pascal's Triangle
    return pascalTriangle;
}

File: 0118-pascals-triangle/0118-pascals-triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        """
        Generate the first 'numRows' of Pascal's Triangle.

        Parameters:
        numRows (int): The number of rows of Pascal's Triangle to generate.

        Returns:
        List[List[int]]: A list of lists, where each inner list represents a row of Pascal's Triangle.
        """

        # Special case: If numRows is 1, return just the first row
        if numRows == 1:
            return [[1]]

        # Special case: If numRows is 2, return the first two rows
        if numRows == 2:
            return [[1], [1, 1]]

        # Initialize Pascal's Triangle with the first two rows
        pascal_triangle = [[1], [1, 1]]

        # Generate rows from the 3rd row onwards (index 2)
        for i in range(2, numRows):
            row = []  # Initialize the current row

            # Generate each value in the current row
            for j in range(i + 1):
                # The first and last values in each row are always 1
                if j == 0 or j == i:
                    row.append(1)
                else:
                    # Each value is the sum of the two values directly above it from the previous row
                    row.append(pascal_triangle[i - 1][j - 1] + pascal_triangle[i - 1][j])

            # Add the generated row to Pascal's Triangle
            pascal_triangle.append(row)

        # Return the full Pascal's Triangle
        return pascal_triangle
